File: q1-25/q16.py
# 2020.07.20
# Problem Statement:
# https://leetcode.com/problems/3sum-closest/

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        
        # the idea is same as question 15, but the implementation is much simpler
        nums.sort()
        
        # initialize current smallest difference between sum of 3 numbers and target number
        current_smallest_diff = inf
        
        # fix one element
        for i in range(0, len(nums)-2):
            left = i+1
            right = len(nums)-1
            
            # loop when left < right
            while left < right:
                # calculate the sum
                sum_3 = nums[i] + nums[left] + nums[right]
                # update the current best result
                if abs(sum_3 - target) < current_smallest_diff:
                    current_smallest_diff = abs(sum_3 - target)
                    answer = sum_3

                # move closer to target    
                if sum_3 < target:
                    left = left + 1
                elif sum_3 > target:
                    right = right - 1
                # if equals to target, directly return    
                else:
                    return target
                
        return answer
            

File: Amazon/q16.java
// 2022.05.23
// Problem Statement:
// https://leetcode.com/problems/3sum-closest/

// idea: sort and fix one, then move left and right
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int diff = Integer.MAX_VALUE;
        int answer = 0;
        for (int i=0; i<nums.length-2; i++) {
            int fixed = nums[i];
            int left = i+1;
            int right = nums.length-1;
            while (left<right) {
                if (nums[left]+nums[right]+fixed==target) return target;
                else if (nums[left]+nums[right]+fixed < target) {
                    if (Math.abs(nums[left]+nums[right]+fixed-target) < diff) {
                        diff = Math.abs(nums[left]+nums[right]+fixed-target);
                        answer = nums[left]+nums[right]+fixed;
                    }
                    left++;
                } else {
                    if (Math.abs(nums[left]+nums[right]+fixed-target) < diff) {
                        diff = Math.abs(nums[left]+nums[right]+fixed-target);
                        answer = nums[left]+nums[right]+fixed;
                    }
                    right--;
                }
            }
            
        }
        return answer;
    }
}